Perf c2c dev

jiebinn opened the pull request:
<!-- CURSOR_SUMMARY -->
> [!NOTE]
> Introduces a new symbol-level C2C TUI and refactors shared types into a header.
> 
> - Adds `ui/browsers/c2c-symbol.c`: a 3-level symbol browser (symbol → related symbol → shared cacheline), sorted by `cycles_percent`, with expand/collapse and cacheline detail navigation
> - Exposes APIs and state: moves C2C structs to new `c2c.h`, makes `c2c` global, and exports `perf_c2c__browse_cacheline()`; declares `perf_c2c__browse_symbol_view()`
> - Integrates with existing C2C browser: adds TAB to switch to Symbol view and updates help text
> - Updates build to compile `c2c-symbol.o`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 0459c2ddbf905ac920534f6e4fd90394ebb36eb0. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

jiebinn committed:
new feature can work
rebase master

jiebinn committed:
metrics

jiebinn committed:
remove redundant code

jiebinn committed:
polish the code

jiebinn committed:
add code address and correct the metrics

jiebinn committed:
modify the title

jiebinn committed:
put cycles_percent first

jiebinn committed:
add code address for related symbols

jiebinn committed:
add metrics for sub symbols

jiebinn committed:
only show several metrics

jiebinn committed:
add metrics in sub symbol cacheline view

jiebinn committed:
modify the store metrics

jiebinn committed:
revise the title

jiebinn committed:
modify the cacheline view similar to `Shared Data Symbols Table`

jiebinn committed:
revise the title of cacheline view

jiebinn committed:
Ensure detail view uses original cacheline details

jiebinn committed:
fix a bug

jiebinn committed:
add the total store metric

jiebinn committed:
iaddr as the key for parent and child relation

jiebinn committed:
fix

jiebinn committed:
revise the metric of cacheline view

jiebinn committed:
fix the store refs

jiebinn committed:
child symbol sorted by Store Refs

jiebinn committed:
rewrite the cacheline view

jiebinn committed:
adjust the symbol view

jiebinn committed:
remove store refs

jiebinn committed:
keep the 90% cacheline view the same as before

jiebinn committed:
keep the cacheline view the same as original

jiebinn committed:
sort by Off

jiebinn committed:
adjust the symbol view

jiebinn committed:
ui optimization

jiebinn committed:
remove redundant code

jiebinn committed:
remove debug code

jiebinn committed:
remove extra indent

jiebinn committed:
remove redundant code

jiebinn committed:
polish the code

jiebinn committed:
polish the code

jiebinn committed:
remove store refs

jiebinn committed:
remove redundant code

jiebinn committed:
restore code

jiebinn committed:
format

jiebinn committed:
fix the bug

jiebinn committed:
rebase master and fix conflict

jiebinn committed:
polish the code

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -914,6 +1069,44 @@ PERCENT_FN(st_l1hit)
+{
+	int width = c2c_width(fmt, hpp, he->hists);
+	double per = percent(he_stats(he)->st_l1hit, total_stats(he)->st_l1hit);
+	char buf[10];
```
if 10 char enough for double? any out of buffer risk?

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -1179,12 +1372,11 @@ node_entry(struct perf_hpp_fmt *fmt __maybe_unused, struct perf_hpp *hpp,
 	bool first = true;
 	int node;
 	int ret = 0;
+	DECLARE_BITMAP(set, c2c.cpus_cnt);
```
why we have this change?

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -1560,15 +1753,15 @@ static struct c2c_dimension dim_ld_fbhit = {
 
 static struct c2c_dimension dim_ld_l1hit = {
-	.header		= HEADER_SPAN_LOW("L1"),
+	.header		= HEADER_SPAN_LOW("L1 Load"),
```
It's better not to change original header.

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -1773,6 +2052,199 @@ static struct c2c_dimension dim_mean_lcl_peer = {
+
+        percent = denom_cycles ? (double)symbol_cycles / denom_cycles * 100.0 : 0.0;
+    } else {
+        total_cycles = get_total_cycles_all_symbols();
```
How many times get_total_cycles_all_symbols() will be called during perf c2c report?

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -1773,6 +2052,199 @@ static struct c2c_dimension dim_mean_lcl_peer = {
+	struct c2c_hist_entry *c2c_right = container_of(right, struct c2c_hist_entry, he);
+	uint64_t cycles_left, cycles_right;
+
+	cycles_left = calculate_symbol_total_cycles(c2c_left);
+	cycles_right = calculate_symbol_total_cycles(c2c_right);
```
to speedup, can we store total cycles as an attribute to hist_entry instread of doing calculation each time.

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -2547,9 +3124,804 @@ static void perf_c2c__hists_fprintf(FILE *out, struct perf_session *session)
+            struct hist_entry *child_he;
+            rel_sym = sorted[i];
+
+		if (!rel_sym || !rel_sym->sym) {
+			continue;
+        }
```
seems there is indent mismatch, try to format with clang-format:
https://www.kernel.org/doc/html/next/dev-tools/clang-format.html

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -2547,9 +3124,804 @@ static void perf_c2c__hists_fprintf(FILE *out, struct perf_session *session)
+        }
+
+        /* Allocate child hist_entry - simplified version for symbol children */
+		child_c2c_he = zalloc(sizeof(*child_c2c_he));
```
why here use zalloc, but Line 3196 use calloc?

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -2547,9 +3124,804 @@ static void perf_c2c__hists_fprintf(FILE *out, struct perf_session *session)
+
+	/* Search for existing entry */
+	for (entry = symbol_hash[hash]; entry; entry = entry->next) {
+		if (entry->iaddr == iaddr && entry->sym == sym) {
```
what's the primary type of `sym`, 
for `entry->sym == sym`, is the pointer cmp, or string cmp?

jiebinn committed:
remove redundant code

jiebinn committed:
cache symbol_total_cycles as global var

jiebinn committed:
cache the total cycles for each symbol

jiebinn committed:
fix a bug to use string cmp

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -3333,11 +3333,13 @@ static void populate_symbol_children(struct hist_entry *he)
 						struct hist_entry *he_d = rb_entry(nd_d, struct hist_entry, rb_node);
 						u64 iaddr_d = he_d->mem_info ? mem_info__iaddr(he_d->mem_info)->addr : 0;
-						if (he_d->ms.sym == he->ms.sym) {
+						if (he_d->ms.sym != NULL && he->ms.sym != NULL && 
+						    strcmp(he_d->ms.sym->name, he->ms.sym->name) == 0) {
```
if this is using so frequently, why not wrap it as an inline function?

jiebinn committed:
inline the strcmp of symbols

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -171,16 +223,67 @@ static void *c2c_he_zalloc(size_t size)
+
+		if (child_he->stat_acc) {
+			free(child_he->stat_acc);
+			child_he->stat_acc = NULL;
```
`zfree()` is doing the same thing.

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -2547,9 +3139,814 @@ static void perf_c2c__hists_fprintf(FILE *out, struct perf_session *session)
+                sorted[i] = sorted[max_j];
+                sorted[max_j] = tmp;
+            }
+        }
```
linux kernel has sort lib: `<linux/sort.h>` and list sort `<linux/list_sort.h>

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -2547,9 +3139,814 @@ static void perf_c2c__hists_fprintf(FILE *out, struct perf_session *session)
+			struct rb_node *nd_cl = rb_first_cached(&c2c.hists.hists.entries);
+			struct rb_root_cached *groot = &child_he->hroot_out;
+			/* temp array to sort grandchildren by Stores */
+			struct grand_item { struct c2c_hist_entry *grand_c2c; struct hist_entry *grand_he; u64 stores; } *items = NULL;
```
struct definition should using multiple lines.

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -2547,9 +3139,814 @@ static void perf_c2c__hists_fprintf(FILE *out, struct perf_session *session)
+					/* push into temp array */
+					if (items_cnt == items_cap) {
+						int new_cap = items_cap ? items_cap * 2 : 8;
+						struct grand_item *ni = realloc(items, new_cap * sizeof(*items));
```
here we allocate the `*item` memory, it's a pointer array to `struct grand_item`, but where is the allocation of struct itself?

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -2547,9 +3139,814 @@ static void perf_c2c__hists_fprintf(FILE *out, struct perf_session *session)
+        struct related_symbol **sorted = NULL;
+
+        /* Count related symbols and allocate array in one pass */
+        list_for_each_entry(rel_sym, &c2c_he->related_symbols, list)
```
this function seems using space as indent

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -2547,9 +3139,814 @@ static void perf_c2c__hists_fprintf(FILE *out, struct perf_session *session)
+ * This enables the hierarchical display in the TUI browser
+ * Children are inserted sorted by Stores (stats.store) descending
+ */
+static void populate_symbol_children(struct hist_entry *he)
```
the function is too long, should split into smaller function to improve readability.

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -2547,9 +3139,814 @@ static void perf_c2c__hists_fprintf(FILE *out, struct perf_session *session)
+
+/* Simple hash table for symbol aggregation */
+#define SYMBOL_HASH_SIZE 1024
+static struct symbol_entry *symbol_hash[SYMBOL_HASH_SIZE];
```
1. is 1024 size enough?
2. why create a big static table?

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -2547,9 +3139,814 @@ static void perf_c2c__hists_fprintf(FILE *out, struct perf_session *session)
+
+static unsigned int symbol_hash_func(uint64_t iaddr, struct symbol *sym)
+{
+	return (iaddr ^ (unsigned long)sym) % SYMBOL_HASH_SIZE;
```
here we use sym addr as hash key, is the same sym name always point to the same *sym.

guowangy commented on tools/perf/builtin-c2c.c:
```diff
@@ -2547,9 +3139,814 @@ static void perf_c2c__hists_fprintf(FILE *out, struct perf_session *session)
+	entry->map = map;
+	entry->maps = maps;
+	memset(&entry->stats, 0, sizeof(entry->stats));
+	memset(&entry->cstats, 0, sizeof(entry->cstats));
```
zalloc already memset

guowangy commented:
can we do a round of perf profiling of c2c report to identify which part needs performance optimization?

jiebinn committed:
revise code based on wangyang review

jiebinn committed:
split populate_symbol_children into several smaller funtions

jiebinn committed:
use cache to avoid multiple call of rb_next

jiebinn committed:
polish the code

jiebinn committed:
polish the code

jiebinn committed:
remove redundant hash table for cache

jiebinn committed:
remove redundant grand child cacheline details

jiebinn committed:
polish the code

jiebinn committed:
code style check

jiebinn committed:
code style polish

jiebinn committed:
code style polish

jiebinn committed:
polish the code

jiebinn committed:
code style polish

jiebinn commented:
@cursor review the PR

jiebinn committed:
fix a bug

jiebinn committed:
split the symbol view

jiebinn committed:
free memory

jiebinn committed:
fix the memory leak issue

jiebinn committed:
refactor code

jiebinn committed:
move symbol view related code to c2c-symbol.c

jiebinn committed:
fix memory leak issue if allocate failure

jiebinn committed:
refator the code

jiebinn committed:
refactor the code

jiebinn committed:
refactor the code

jiebinn committed:
refactor the code

jiebinn committed:
polish the code

jiebinn committed:
add check

jiebinn commented on tools/perf/builtin-c2c.c:
```diff
@@ -914,6 +857,47 @@ PERCENT_FN(st_l1hit)
 PERCENT_FN(st_l1miss)
 PERCENT_FN(st_na)
 
+/* Simple percentage functions for cacheline offsets view */
```
move the symbol view entry to c2c-symbol.c

jiebinn commented on tools/perf/builtin-c2c.c:
```diff
@@ -1781,7 +1807,96 @@ static struct c2c_dimension dim_cpucnt = {
 };
 
-static struct c2c_dimension dim_srcline = {
+/* New dimensions for latency and cycles calculations with English naming */
```
code clean up

jiebinn commented on tools/perf/builtin-c2c.c:
```diff
@@ -1834,6 +1949,9 @@ static struct c2c_dimension *dimensions[] = {
 	&dim_cl_stores_l1hit,
 	&dim_cl_stores_l1miss,
 	&dim_cl_stores_na,
```
maintain symbol view in another file.

jiebinn commented on tools/perf/builtin-c2c.c:
```diff
@@ -2690,44 +2825,91 @@ perf_c2c_browser__new(struct hists *hists)
 	return browser;
 }
 
```
only one hot key to jump to a new browser.

jiebinn commented on tools/perf/builtin-c2c.c:
```diff
@@ -162,6 +102,13 @@ static void *c2c_he_zalloc(size_t size)
+
+	/* Initialize cached total cycles */
+	c2c_he->total_cycles = 0;
+	c2c_he->total_cycles_valid = false;
```
consider where to add these global? or in a seperate funtion.

jiebinn committed:
clean up code

jiebinn committed:
clean up code

jiebinn committed:
fix a TUI bug

jiebinn committed:
refactor the tui code

jiebinn committed:
optimize the potential overflow issue

jiebinn committed:
add fast path in iaddr_entry

jiebinn committed:
pack the init/clean function

jiebinn commented on tools/perf/builtin-c2c.c:
```diff
@@ -174,8 +117,13 @@ static void *c2c_he_zalloc(size_t size)
 	c2c_he = container_of(he, struct c2c_hist_entry, he);
+
+	/* Clean up symbol view support */
+	c2c_he_cleanup_symbol_support(hist_entry, c2c_he);
```
rename

jiebinn committed:
optimize header file, remove redundant header. Rename functions with reasonable name

jiebinn committed:
decouple cacheline browser and symbol browser

jiebinn committed:
avoid recursive call of browser

jiebinn committed:
decouple dim_iaddr_symbol

jiebinn committed:
refact the c2c_hist_entry

jiebinn committed:
add perf_c2c_ext

jiebinn committed:
decouple dim symbol in two views

jiebinn committed:
decouple dimensions

jiebinn committed:
perf c2c: Fix symbol view allocation using uninitialized CPU/node counts

The c2c_he_ext_zalloc function was incorrectly using c2c_ext.c2c.cpus_cnt
and c2c_ext.c2c.nodes_cnt, which are never initialized (remain zero).
These should be c2c.cpus_cnt and c2c.nodes_cnt, which are properly
initialized during analysis.

This caused bitmap_zalloc(0) and zalloc(0) calls, leading to symbol
view histogram entry allocation failures.

jiebinn committed:
decouple init and reinit function in two views

jiebinn committed:
add static

jiebinn committed:
remove redundant code

jiebinn committed:
polish the code

jiebinn committed:
polish the code

jiebinn committed:
polish the code

jiebinn committed:
remove cache index clear

jiebinn committed:
add basic c2c_dimension in c2c symbol

jiebinn committed:
remove perf_c2c from perf_c2c_ext

jiebinn committed:
remove redundant code

jiebinn committed:
refactor the code to remove c2c_hist_entry_ext

jiebinn committed:
refacor the code to remove *sym

jiebinn committed:
add _symbol_accessed_cachelines

jiebinn committed:
refactor to remove temp struct

jiebinn committed:
merge step1, 2, 4

jiebinn committed:
remove child_cacheline_ref

jiebinn committed:
remove symbol_agg_entry

jiebinn committed:
rename function

jiebinn committed:
refactor the code with basic struct
